roachtest: add asyncpg test suite

ZhouXing19 opened the pull request:
Fix #70209.

Add a new roachtest, asyncpg, that changes the connection port
to point at cockroachdb, and runs the original test suite from
the asyncpg package (`python3 setup.py test`).

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71029)
<!-- Reviewable:end -->

ZhouXing19 commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1737,6 +1737,26 @@ fi
+// GitCloneWithRecurseSubmodules clones a git repo from src into dest and checks out origin's
+// version of the given branch, but with a --recurse-submodules flag.
+// The src, dest, and branch arguments must not contain shell special characters.
+func (c *clusterImpl) GitCloneWithRecurseSubmodules(
```
@tbg TFTR! Moved it to `tests/canary.go`.

otan commented on pkg/cmd/roachtest/tests/asyncpg.go:
```diff
@@ -0,0 +1,114 @@
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+)
+
+func registerAsyncpg(r registry.Registry) {
```
nit: to make this Go style compliant, we should call everything `AsyncPG` (acronyms are always the same case, capital letters if it is not the first word, e.g. pgCompat, compatPG)

otan replied:
hmm, actually maybe ignore this -- it's stylised asyncpg /shrug

otan commented on pkg/cmd/roachtest/tests/asyncpg.go:
```diff
@@ -0,0 +1,114 @@
+			c,
+			node,
+			"install postgresql",
+			`sudo apt-get -y install postgresql`,
```
curious, why do we need this? can you add a comment?

otan commented on pkg/cmd/roachtest/tests/canary.go:
```diff
@@ -291,3 +292,27 @@ func repeatGetLatestTag(
+  git fetch origin
+  git checkout origin/%s
+fi
+'`, dest,
```
let's use `%[1]s, %[2]s, ...` here to save duplicate arguments.

```
`if ! test -d %[1]s; then
  git clone --recurse-submodules -b %[2]s --depth 1 %[3]s %[1]s
else
  cd %[1]s
  git fetch origin
  git checkout origin/%[2]s
fi`,
dest,
branch,
src,
```

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @tbg, and @ZhouXing19)

___
*[pkg/cmd/roachtest/tests/asyncpg.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71029#-MkzFG1HtDNRmlUjfPR6-r2-77:-MlL34P60eeYaok0jAyg:bp36pvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/105cd1261f19a360db9e68cbafa51e093a333ec6/pkg/cmd/roachtest/tests/asyncpg.go#L77)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

curious, why do we need this? can you add a comment?
</blockquote></details>

It's because asyncpg requires a `pg_ctl` for the setup. See [asyncpg's code here](https://github.com/MagicStack/asyncpg/blob/383c711eb68bc6a042c121e1fddfde0cdefb8068/asyncpg/cluster.py#L407-L420).

Without this PostgreSQL installation, `roachtest run asyncpg` gives:
```
Traceback (most recent call last):
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 319, in setUpClass
    cls.setup_cluster()
  File "/mnt/data1/asyncpg/tests/test_connect.py", line 1222, in setup_cluster
    cls.cluster = cls.new_cluster(pg_cluster.TempCluster)
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 298, in new_cluster
    cluster = _init_cluster(ClusterCls, cluster_kwargs,
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 211, in _init_cluster
    cluster.init(**(initdb_options or {}))
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 109, in init
    if self.get_status() != 'not-initialized':
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 77, in get_status
    self._init_env()
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 417, in _init_env
    self._pg_ctl = self._find_pg_binary('pg_ctl')
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 554, in _find_pg_binary
    raise ClusterError(
asyncpg.cluster.ClusterError: could not find pg_ctl executable: '/usr/lib/postgresql/12/bin/pg_ctl' does not exist or is not a file
```
___
*[pkg/cmd/roachtest/tests/canary.go, line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71029#-MlAvUmWTw57ZaMUb0ri-r2-314:-MlL2ZTW4tZte1tBDr3p:b70mqqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/105cd1261f19a360db9e68cbafa51e093a333ec6/pkg/cmd/roachtest/tests/canary.go#L314)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

let's use `%[1]s, %[2]s, ...` here to save duplicate arguments.

```
`if ! test -d %[1]s; then
  git clone --recurse-submodules -b %[2]s --depth 1 %[3]s %[1]s
else
  cd %[1]s
  git fetch origin
  git checkout origin/%[2]s
fi`,
dest,
branch,
src,
```
</blockquote></details>

Thanks! TIL.


<!-- Sent from Reviewable.io -->

otan reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @tbg, and @ZhouXing19)

___
*[pkg/cmd/roachtest/tests/asyncpg.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71029#-MkzFG1HtDNRmlUjfPR6-r2-77:-MlgdfuJ2zDh-SD9jxR5:b-dik1fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/105cd1261f19a360db9e68cbafa51e093a333ec6/pkg/cmd/roachtest/tests/asyncpg.go#L77)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

It's because asyncpg requires a `pg_ctl` for the setup. See [asyncpg's code here](https://github.com/MagicStack/asyncpg/blob/383c711eb68bc6a042c121e1fddfde0cdefb8068/asyncpg/cluster.py#L407-L420).

Without this PostgreSQL installation, `roachtest run asyncpg` gives:
```
Traceback (most recent call last):
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 319, in setUpClass
    cls.setup_cluster()
  File "/mnt/data1/asyncpg/tests/test_connect.py", line 1222, in setup_cluster
    cls.cluster = cls.new_cluster(pg_cluster.TempCluster)
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 298, in new_cluster
    cluster = _init_cluster(ClusterCls, cluster_kwargs,
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 211, in _init_cluster
    cluster.init(**(initdb_options or {}))
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 109, in init
    if self.get_status() != 'not-initialized':
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 77, in get_status
    self._init_env()
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 417, in _init_env
    self._pg_ctl = self._find_pg_binary('pg_ctl')
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 554, in _find_pg_binary
    raise ClusterError(
asyncpg.cluster.ClusterError: could not find pg_ctl executable: '/usr/lib/postgresql/12/bin/pg_ctl' does not exist or is not a file
```
</blockquote></details>

hmm, not a big fan of it starting postgres to execute its tests. i wish we could comment it out. oh well.
it needs pg_ctl AND postgres to setup from what I can read. Can you link the code snippet in your comment?


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! but needs one more `make bazel-generate`

ZhouXing19 committed:
roachtest: add asyncpg test suite

Fix #70209.

Add a new roachtest, asyncpg, that changes the connection port
to point at cockroachdb, and run the original test suite from
the asyncpg package (`python3 setup.py test`).

Release note: None

ZhouXing19 commented:
bors r+
